core/primitives/src/time.rs

```diff
@@ -1,155 +1,300 @@
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+    impl BorshSerialize for Time {
+        fn serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
+            let nanos = self.to_unix_timestamp_nanos().as_nanos() as u64;
+            BorshSerialize::serialize(&nanos, writer).unwrap();
+            Ok(())
+        }
     }
 
-    pub fn add_instant(mock_instant: Duration) {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                clock.durations.push_back(mock_instant);
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+    impl BorshDeserialize for Time {
+        fn deserialize(buf: &mut &[u8]) -> Result<Self, std::io::Error> {
+            let nanos: u64 = borsh::BorshDeserialize::deserialize(buf)?;
+
+            Ok(Time::from_unix_timestamp(Duration::from_nanos(nanos)))
+        }
     }
 
-    pub fn utc() -> DateTime<chrono::Utc> {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                let x = clock.pop_utc();
-                match x {
-                    Some(t) => t,
-                    None => {
-                        panic!("Mock clock run out of samples");
-                    }
-                }
-            } else {
-                chrono::Utc::now()
-            }
-        })
+    impl From<SystemTime> for Time {
+        fn from(system_time: SystemTime) -> Self {
+            Self { system_time }
+        }
     }
 
-    pub fn instant() -> Instant {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                let x = clock.pop_instant();
-                match x {
-                    Some(t) => t,
-                    None => {
-                        panic!("Mock clock run out of samples");
-                    }
-                }
-            } else {
-                Instant::now()
-            }
-        })
+    impl From<DateTime<Utc>> for Time {
+        fn from(utc: DateTime<Utc>) -> Self {
+            // utc.timestamp_nanos() returns i64
+            let nanos = utc.timestamp_nanos() as u64;
+
+            Self::UNIX_EPOCH + Duration::from_nanos(nanos)
+        }
+    }
+
+    impl Time {
+        pub const UNIX_EPOCH: Time = Time { system_time: SystemTime::UNIX_EPOCH };
+
+        pub fn now() -> Self {
+            Self::UNIX_EPOCH + Duration::from_nanos(Clock::utc().timestamp_nanos() as u64)
+        }
+
+        pub fn duration_since(&self, rhs: &Self) -> Duration {
+            self.system_time.duration_since(rhs.system_time).unwrap_or(Duration::from_millis(0))
+        }
+
+        pub fn elapsed(&self) -> Duration {
+            Self::now().duration_since(self)
+        }
+
+        pub fn from_unix_timestamp(duration: Duration) -> Self {
+            Self::UNIX_EPOCH + duration
+        }
+
+        pub fn to_unix_timestamp_nanos(&self) -> Duration {
```

Comment:
This "nanos" in the naming does not match my expectations given the `Duration` return value. To be honest, "unix_timestamp" as Duration is also a bit misleading (should we create a wrapper type or at least a type-alias?)

Reply:
> This "nanos" in the naming does not match my expectations given the `Duration` return value. 

You are right, I missed that.

> To be honest, "unix_timestamp" as Duration is also a bit misleading

That's a good point. I'll rename it

>  (should we create a wrapper type or at least a type-alias?)

Yes, I think we should. I'll do it

Reply:
@frol @matklad Should we use 
```
type UnixTimeStamp = Duration;
```
or is it better to use keep u64 in those methods for compatibility reasons?

Reply:
> I am fine with the implementation, but I would keep this in a PR state until we actually add some usage to it, so we can see if we need anything else added to it.

Ok, I'll wait until I rewrite the other PR which adds support of this to `near-network`.

Reply:
After thinking about it:
```
type UnixTimeStamp = u64
```
Makes most sense. We only use nanos as UnixTimeStamp, there is no point in using duration.